### ISSUE TITLE
[Question] About alternatives to tokenizations

### ISSUE THREAD

**Issue by loretoparisi** [polm/fugashi#23]
Thanks for your post about how to tokenize Japanese.
Currently my solution is to use `icu` tokenizer with word break iterator and customized locale as showed here:

![code](https://user-images.githubusercontent.com/163333/90143131-862d7d00-dd7d-11ea-8359-ae01388d0d0f.png)
[Repl.it](https://repl.it/@loretoparisi/JapaneseTokenizer#icu.py)
My question is if this approach has in any cases the same results as `fugashi` for what concerns the japanese language.

Thank you!

**Comment by polm** [comment]
So that's an excellent question. 

The issue with ICU is that it's (obviously) not Japanese specific, and it's kind of weird it includes this functionality at all. Its tokenizer isn't completely awful, and for the example text in this case it gives exactly the same results as fugashi with unidic-lite, but there are other cases where it will be different or just bad. 

The way MeCab works is it builds a lattice and uses the Viterbi algorithm to find the cheapest path through the lattice to determine how to split a sentence. This uses both the cost of individual words (common words are cheaper) and the cost of part-of-speech transitions. 

ICU has a surprisingly similar method, but it doesn't have part of speech information. The dictionary it uses is IPAdic (which you can also use with fugashi, but shouldn't, [see here](https://github.com/polm/ipadic-py)), plus one (or more?) Chinese dictionaries. 

If you look at the ICU dictionary file for JA/ZH, it has costs for words, and it does actually calculate costs across a sentence. But without part of speech information there are limits to what it can do. (Including Chinese words also probably causes some problems.) Here's a sentence where it breaks down:

    すももももももももの内
    correct: すもも も もも も もも の 内
    ICU: すもも も も も も も もの 内

What happened here is that the particle も has a very low cost because it's a fundamental grammatical word, and without part of speech information ICU just picks it repeatedly. 

So, to sum up:

1. the ICU breaks aren't total garbage, but
2. you can't reproduce them exactly in fugashi all the time
3. when different, they'll usually be worse than fugashi
4. they are based on IPAdic and will be closer to fugashi+IPAdic than fugashi+UniDic

Let me know if that answers your question.

**Comment by loretoparisi** [comment]
@polm thanks a lot for the detailed explanation. I was not aware of the `POS` missing in ICU. Since I'm using MeCab for other things, and I have recently switched to the latest `mecab==1.0.1` with `unidic-lite` I will try then [fugashi](url) as alternative at least when `lang==ja`.

**Comment by polm** [comment]
Great, glad it was helpful! 

Note that fugashi and mecab-python3 are both just wrappers for MeCab, so if the dictionary is the same the output should also be the same. Fugashi is just more Pythonic, so you can get named tuples with info etc. 

Also note that `mecab` on pypi is a separate project not maintained by me, but should also give the same output with the same dictionary.

**Comment by loretoparisi** [comment]
@polm thanks, so just to clarify, I was previously using `MeCab` as tokenizer for another project in this way:

```python
mtok = MosesTokenizer(lang = lang)
line = mpn.normalize(line)
tok = mtok.tokenize(line, escape=False, return_str=True, aggressive_dash_splits=False, protected_patterns=patterns)
mecab = MeCab.Tagger(ipadic.MECAB_ARGS + " -Owakati -b 5000")
tokens = mecab.parse(tok)
```

so using the `MosesTokenizer` and then MeCab segmenter, could I safely replace this pipeline for `lang==ja` with `fugashi`?

Thanks!

**Comment by polm** [comment]
If you are using `MeCab.parse` from `mecab-python3` then fugashi is exactly the same, since the underlying library is MeCab. The main advantage is if you need data like lemmas, fugashi will parse the dictionary data into a named tuple for easier access.

Note that `mecab` on PyPI should also be the same, but I don't maintain that project so I'm not 100% sure.

I don't really know what Moses does beyond tokenizing English etc., but if you have fugashi/MeCab I don't think you need it.

**Comment by loretoparisi** [comment]
@polm okay thanks a lot, I'm going to replace to `fugashi` this japanese pipeline then. Keep open this issue while I come back with results of this port from ICU / MeCab (base) to your package.

Thank you.

**Comment by polm** [comment]
Since it's been a month now I'm going to go ahead and close this, but please feel free to comment/reopen it if you have any updates.
